Frameworks/include/CFCppBase.h

```diff
@@ -0,0 +1,90 @@
+//******************************************************************************
+//
+// Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#pragma once
+
+#ifndef __cplusplus
+#pragma error CFCppBase is designed for C++ consumers only.
+#endif
+
+#include <CFRuntime.h>
+
+#include <type_traits>
+#include <typeinfo>
+#include <new>
+
+namespace CoreFoundation {
+namespace Details {
+struct Nothing {};
+
+template <typename P>
+struct CppBaseImpl : P {};
+
+template <>
+struct CppBaseImpl<Details::Nothing> : __CFRuntimeBase {
+    static inline void _CFInit(CFTypeRef cf) {
+    }
+    static inline void _CFFinalize(CFTypeRef cf) {
+    }
+};
+}
+
+template <typename T, typename TImpl = typename T::Impl, typename Parent = Details::Nothing>
+struct CppBase : Details::CppBaseImpl<Parent> {
```

Comment:
Your usage in CGContext.mm is self explanatory.  Would be awesome to have some comments here that show the usage should someone need it.